### PR TITLE
[WIP] Inventory groups

### PR DESCRIPTION
**Pull request by sulyi**
## Summary ##

SUMMARY: Interface "Add inventory groups"

## Purpose of change ##

Collapsible and expandable inventory entries, containing other entries.
 - Supporting interactions like dropping, favouring, etc ...
 - Having multiple parents containing them.
 - Showing number of stacks and total charge in them.

## Describe the solution ##

Many features are missing, such as:
 - [x] group entries are created in `inventory_column::prepare_paging`
 - [x] collapsed and expanded state both works
 - [ ] charge calculation (currently only charge of the first item counts)
 - [ ] indentation (currently groups drawn with same indent as items)
 - [ ] ~~assigning intvlet~~ (needs to be forbidden for groups)
 - [x] favouriting
 - [ ] dropping
   - [x] group of stacks
   - [ ] group of items with charged (no. of stacks in group is dropped from first stack)
 - [ ] cleanup (infrastructure is ready)
 - [ ] group state (~~currently groups can only be collapsed/expanded together~~, toggle is missing)

## Testing ##

There are few testcases, but more would be nice (suggestions are well come):
 - [ ] group end of inventory column
 - [ ] group of items with charge (marijuana, rolling parer)
 - [ ] items with max charge (flashlight, lighter)
 - [x] group of items already in a stack (toothbrush)
 - [ ] containment indicators
 - [ ] items with damage
 - [x] single item in inventory
 - [ ] filtering
 - [ ] liquids

_Note:_ groups are made by putting items in different containers (e.g. skirt, trenchcoat, backpack).

## TL;DR ##

Here are some screenshots of the  current(ish) state:

<table width="100%">
<tr><td><b>expanded:</b></td><td><b>collapsed:</b></td></tr>
<tr>
<td valign="top">

</td>
<td valign="top">

</td>
</tr>
</table>

_Note:_ in both picture inventory has: 
 - in skirt: marijuana (4), rolling papers (23), 3 toothbrush (filthy), 
 - in coat marijuana (1), rolling paper (1), 2 toothbrushes (filthy),
 - in backpack: rolling paper (5).
 - I forgot, where rolling tobacco (15) and rolling tobacco (5) is, but one of them in the plastic bag (some how, it's not showing)

### CODE REVIEW

File: src/inventory_ui.h [hunk 1]
```diff
@@ -148,6 +152,44 @@ class inventory_entry
...
+         */
+        bool is_item() const {
+            return false;
+        }
+
+        int get_number_of_children() const {
+            return n_children;
+        }
+
+        void set_number_of_children( const int n );
```
Comment (BrettDong) [inline src/inventory_ui.h:187]:
```suggestion
        void set_number_of_children( const int &n );
```

File: src/inventory_ui.cpp [hunk 2]
```diff
@@ -218,6 +218,11 @@ inventory_entry *inventory_column::find_by_invlet( int invlet ) const
     return nullptr;
 }
 
+void group_entry::set_number_of_children( const int n )
```
Comment (BrettDong) [inline src/inventory_ui.cpp:221]:
```suggestion
void group_entry::set_number_of_children( const int &n )
```

Comment (anothersimulacrum) [reply src/inventory_ui.cpp:221]: It's fine to (maybe even cheaper to) copy ints, the function definition should just lose the const.

File: src/inventory_ui.h (hunk 1 above)
Comment (anothersimulacrum) [reply src/inventory_ui.h:187]:
```suggestion
        void set_number_of_children( int n );
```

File: src/inventory_ui.cpp (hunk 2 above)
Comment (BrettDong) [reply src/inventory_ui.cpp]: Isn't reference cheaper?

Comment (sulyi) [reply src/inventory_ui.cpp]: Wasn't thinking much about it, other than needs to be enclosed, but yes it's probably way overkill.

Comment (anothersimulacrum) [reply src/inventory_ui.cpp]: Not for an int.
```cpp
#include <cstdio>
#include <string>

int main()
{
    class intref
    {
        int &ref;
        intref(int a) : ref(a) {}
    };
    class stringref
    {
        std::string &ref;
        stringref(std::string a) : ref(a) {}
    };
    printf("sizeof int %ld\n", sizeof(int));
    printf("sizeof int & %ld\n", sizeof(intref));
    printf("sizeof int * %ld\n", sizeof(int *));

    printf("sizeof std::sting %ld\n", sizeof(std::string));
    printf("sizeof std::string & %ld\n", sizeof(stringref));
    printf("sizeof std::string * %ld\n", sizeof(std::string *));
    return 0;
}
>>>>>
sizeof int 4
sizeof int & 8
sizeof int * 8
sizeof std::sting 32
sizeof std::string & 8
sizeof std::string * 8
```

File: src/inventory_ui.h [hunk 3]
```diff
@@ -148,6 +152,44 @@ class inventory_entry
...
+
+
+        // TODO: override, override, override
+
+        /**
+         * Group entry is not an item.
+         */
+        bool is_item() const {
+            return false;
+        }
```
Comment (sulyi) [inline src/inventory_ui.h:181]: I'm thinking, don't allow it, unless asked for it. But, this throws me into a philosophical infinite loop.

File: src/inventory_ui.cpp [hunk 4]
```diff
@@ -884,9 +892,71 @@ void inventory_column::prepare_paging( const std::string &filter )
...
+        if( !expd ) { // TODO: change on change
+            // pop group
+            std::vector<inventory_entry>::iterator gp =
+                std::find( entries.begin(), entries.end(), *g_iter );
+            gp++;
+            entries.erase( gp, gp + g_iter->get_number_of_children() );
+            // TODO: not yet, implement proper state maitanence for `group_entry` first
+            // g_iter->expanded = false;
+        }
+    }
```
Comment (sulyi) [inline src/inventory_ui.cpp:954]: This caused so many SIGSEGV, that I'm still afraid somehow I messed up it.

**Review by sulyi — COMMENTED** [review]
Let's, make sure these are proper.

File: src/inventory_ui.cpp (hunk 2 above)
Comment (sulyi) [reply src/inventory_ui.cpp]: Hah, on x86 would be a wash, though.

Comment (anothersimulacrum) [reply src/inventory_ui.cpp]: It's such a micro optimization it doesn't matter, but copying is fine for ints and many of the basic types.

Comment (Qrox) [reply src/inventory_ui.cpp]: A copy can be passed directly via a register, while a reference has to be accessed via the memory, so I'd say a copy of a scalar is always cheaper.

File: src/inventory_ui.h [hunk 5]
```diff
@@ -454,6 +496,14 @@ class inventory_column
...
 
         cata::optional<bool> indent_entries_override = cata::nullopt;
+        /** move this if needs to be disclosed not grouped */
+        bool is_grouped() {
+            return grouped && group_entries_.empty();
+        }
+        /** flags if entries have been grouped and group entries have been added */
+        bool grouped = true;
+        /** stores pointers for group entries */
+        std::vector < std::shared_ptr<group_entry> > group_entries_;
```
[the remark this answers is not in the document]
Comment (Qrox) [reply src/inventory_ui.h:506]: As long as it uses smart pointers the memory should be automatically released on deconstrution.

File: src/inventory_ui.cpp (hunk 4 above)
Comment (Qrox) [reply src/inventory_ui.cpp:954]: You might want to sanity-check the variables and print debug messages / bail out early to avoid hard crashes. (For example, check that `g_iter` is not null, `gp` is not `entries.end()`, `std::distance( gp, entries.end() )` is not lower than `g_iter->get_number_of_children()`

File: src/inventory_ui.cpp [hunk 6]
```diff
@@ -884,9 +892,71 @@ void inventory_column::prepare_paging( const std::string &filter )
...
+                // create group entry
+                int children = std::distance( gp, nx );
+                group_entry g_entry( group, it->get_category_ptr() );
+                g_entry.update_cache();
+                g_entry.highlight_as_parent = true;
+                g_entry.set_number_of_children( children );
+                group_entries_.push_back(  // cant't be in main scope?
+                    std::make_shared<group_entry> ( g_entry ) );
+
+                nx = entries.insert( gp, g_entry );
```
Comment (Qrox) [inline src/inventory_ui.cpp:930]: I don't think this will work as intended. `g_entry` will likely just be up-casted to an `inventory_entry` and added to the list, removing any `group_entry` specific members in the process.

**Comment by Qrox** [comment]
Please also check that filtering works when entries are expanded or collapsed, because the inventory menu implements filtering in a quirky way that is prone to error. (To simply put it, normally one generates the unfiltered entries and make a filtered list of references to the entries. In the inventory menu it first generates the entries and copies the results to the unfiltered entries if the later is empty, so if any changes are made to the entries after that, they have to be mirrored to the unfiltered entries.)

File: src/inventory_ui.cpp (hunk 6 above)
Comment (sulyi) [reply src/inventory_ui.cpp:930]: Yeah, but if I'm adding it to the list why would I cast it? The list is `protected`, adding and removing from it is controlled. So, yes overrides in `selection_column` will need to be checked, but other than that I'd prefer inheritance over extension.

File: src/inventory_ui.cpp [hunk 7]
```diff
@@ -839,29 +846,30 @@ void inventory_column::prepare_paging( const std::string &filter )
...
-    filter, [this]( const std::string & filter ) {
+    const std::function<bool( const inventory_entry & )> filter_fn = \
+    filter_from_string<inventory_entry>( filter, [this]( const std::string & filter ) {
         return preset.get_filter( filter );
     } );
 
     // First, remove all non-items
-    const auto new_end = std::remove_if( entries.begin(),
-    entries.end(), [&filter_fn]( const inventory_entry & entry ) {
+    std::vector<inventory_entry>::const_iterator new_end = std::remove_if(
```
[the remark this answers is not in the document]
Comment (sulyi) [reply src/inventory_ui.cpp:855]: I just make a point about removing all of it from touched members. Also trying to make these kind of changes in separate commits.
I will review these before I'm finished.

File: src/inventory_ui.cpp (hunk 6 above)
Comment (Qrox) [reply src/inventory_ui.cpp:930]: > why would I cast it

Maybe there's some confusion, but I was not suggesting that _you_ should cast it. `g_entry` is _implicitly_ cast in your code, and will only behave like a non-inherited `inventory_entry` in the list, which might not be what you intended.

Comment (sulyi) [reply src/inventory_ui.cpp:930]: I get it, but I use `group_entry` only to insert it to `inventory_column::entries`, and that's only exposed to derived classes.

File: src/inventory_ui.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (sulyi) [reply src/inventory_ui.h:506]: Fixed it.

File: src/inventory_ui.cpp (hunk 6 above)
Comment (Qrox) [reply src/inventory_ui.cpp]: If it's intended, an explicit cast and a comment explaining it would be nice.

**Comment by sulyi** [comment]
I thought, I's in the clear making changes in `prepare_paging`.
But, @Qrox, I think, I saw a seg fault from `ui_manager::invalidate`.
Or that might not be related?

File: src/inventory_ui.cpp [hunk 8]
```diff
@@ -884,9 +892,71 @@ void inventory_column::prepare_paging( const std::string &filter )
             from = to;
         }
     }
+
+    if( !grouped ) {
+        // remove group entries
+        for( const std::shared_ptr< group_entry > &g_iter : group_entries_ ) {
+            // TODO: child entries need to be put back!
+            std::remove( entries.begin(), entries.end(), *g_iter );
```
[the remark this answers is not in the document]
Comment (sulyi) [reply src/inventory_ui.cpp]: Big thanks! At first, I was, like, what.

**Comment by Qrox** [comment]
If you provide a crash log I can try to diagnose it for you.

**Comment by Night-Pryanik** [comment]
@sulyi could you please resolve conflicts?

**Comment by sulyi** [comment]
I was working on other things lately, but I haven't gave up on this.
I had problems with long compile time and difficulty with debugging, due to limitation of my setup, which is a bit discouraging for me.

**Comment by sulyi** [comment]
Thx, github, i didn't mean to merge tho.

**Comment by I-am-Erk** [comment]
Hi sulyi, I am going to close this as stale for now, but if you're still around feel free to reopen it.
